runner/runner.go

```diff
@@ -231,60 +264,77 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
 		}
 	}
 
+	responseChannel := make(chan *Response, len(currentCheckList))
+	wg := &sync.WaitGroup{}
+	
 	// main loop to get the checks info.
 	for _, name := range currentCheckList {
-		resp := &Response{
-			name: name,
-		}
+		wg.Add(1)
+		xname := name
+		go func(xname string, responseChannel chan *Response) {
+			defer wg.Done()
+			currentCheck, ok := checkMap[xname]
+			if !ok {
+				combinedResponse.errs["check not found"] = &Response{
+					name: name,
+				}
+				combinedResponse.checkNotFound = true
+				//continue
+				return
+			}
 
-		currentCheck, ok := checkMap[name]
-		if !ok {
-			combinedResponse.errs["check not found"] = resp
-			combinedResponse.checkNotFound = true
-			continue
-		}
+			// find runner for the given role only.
+			if !currentCheck.verifyRole(r.role) {
+				return
+			}
 
-		// find runner for the given role only.
-		if !currentCheck.verifyRole(r.role) {
-			continue
-		}
+			if _, ok := combinedResponse.checks[name]; ok {
+				combinedResponse.errs["duplicate check"] = &Response{
+					name: name,
+				}
+				return
+			}
 
-		var (
-			combinedOutput []byte
-			code           int
-			err            error
-			checkDuration  string
-		)
+			resp := &Response{
+				name: name,
+			}
 
-		if _, ok := combinedResponse.checks[name]; ok {
-			combinedResponse.errs["duplicate check"] = resp
-			continue
-		}
+			var (
+				combinedOutput []byte
+				code           int
+				err            error
+				checkDuration  string
+			)
 
-		// list option disables the check execution
-		if !list {
 
-			start := time.Now()
-			combinedOutput, code, err = currentCheck.Run(ctx, r.role)
-			checkDuration = time.Since(start).String()
-		}
+			// list option disables the check execution
+			if !list {
 
-		resp.output = string(combinedOutput)
-		resp.status = code
-		resp.duration = checkDuration
-		resp.description = currentCheck.Description
-		resp.cmd = currentCheck.Cmd
-		resp.timeout = currentCheck.Timeout
-		resp.list = list
-		combinedResponse.checks[name] = resp
-
-		// collect errors
-		if err != nil {
-			combinedResponse.errs[err.Error()] = resp
-		}
+				start := time.Now()
+				combinedOutput, code, err = currentCheck.Run(ctx, r.role)
+				checkDuration = time.Since(start).String()
+			}
+
+			resp.output = string(combinedOutput)
+			resp.status = code
+			resp.duration = checkDuration
+			resp.description = currentCheck.Description
+			resp.cmd = currentCheck.Cmd
+			resp.timeout = currentCheck.Timeout
+			resp.list = list
+
+			combinedResponse.checks[name] = resp
+
+			// collect errors
+			if err != nil {
+				combinedResponse.errs[err.Error()] = resp
+			}
 
-		combinedResponse.status = max(combinedResponse.status, code)
+			combinedResponse.status = max(combinedResponse.status, code)
+			
+		}(xname, responseChannel)
 	}
+	wg.Wait()
```

Comment:
What do you think about the following alternative?

```
type responseError struct {
  err error
  checkName string
}

errs := make(chan responseError)
responses := make(chan error)

for _, name := range currentCheckList {
  go func(name string) {
    defer wg.Done()
    currentCheck, ok := checkMap[name]
    if !ok {
      errs <- responseError{errors.New("check not found"), name}
      return
    }
    ...
    // find runner for the given role only.
    if !currentCheck.verifyRole(r.role) {
      // send an empty response for this check, as it is not relevant for this role.
      responses <- nil
      return
    }
    ...
    resp := &Response{ ... }
    ...
    responses <- resp
    return
  }(name)
}

// Expect one response or error for every check
for range currentCheckList {
  select {
  case err := <-errs:
    combinedResponse.errs[err.Error()] = err.name
  case resp := <-responses:
    if resp == nil {
      // skip filtered checks that are not appropriate for this role.
      continue
    }
    combinedResponse.checks[resp.name] = resp
  }
}

combinedResponse.status = ...
return combinedResponse, nil
```

Reply:
This is how I would do it